src/java.base/share/classes/sun/security/rsa/RSAPrivateCrtKeyImpl.java

```diff
@@ -192,20 +193,44 @@ private RSAPrivateCrtKeyImpl(byte[] encoded) throws InvalidKeyException {
         this.keyParams = keyParams;
 
         try {
-            // generate the key encoding
-            DerOutputStream out = new DerOutputStream();
+            byte[][] nbytes = new byte[8][];
+            nbytes[0] = n.toByteArray();
+            nbytes[1] = e.toByteArray();
+            nbytes[2] = d.toByteArray();
+            nbytes[3] = p.toByteArray();
+            nbytes[4] = q.toByteArray();
+            nbytes[5] = pe.toByteArray();
+            nbytes[6] = qe.toByteArray();
+            nbytes[7] = coeff.toByteArray();
+
+            // Initiate with a big enough size so there's no need to
+            // reallocate memory later and thus can be cleaned up
+            // reliably.
+            DerOutputStream out = new DerOutputStream(
+                    nbytes[0].length + nbytes[1].length +
+                    nbytes[2].length + nbytes[3].length +
+                    nbytes[4].length + nbytes[5].length +
+                    nbytes[6].length + nbytes[7].length +
+                    100); // Enough for version(3) and 8 tag+length(3 or 4)
             out.putInteger(0); // version must be 0
-            out.putInteger(n);
-            out.putInteger(e);
-            out.putInteger(d);
-            out.putInteger(p);
-            out.putInteger(q);
-            out.putInteger(pe);
-            out.putInteger(qe);
-            out.putInteger(coeff);
-            DerValue val =
-                new DerValue(DerValue.tag_Sequence, out.toByteArray());
+            out.putInteger(nbytes[0]);
+            out.putInteger(nbytes[1]);
+            out.putInteger(nbytes[2]);
+            out.putInteger(nbytes[3]);
+            out.putInteger(nbytes[4]);
+            out.putInteger(nbytes[5]);
+            out.putInteger(nbytes[6]);
+            out.putInteger(nbytes[7]);
+            // Private values from [2] on.
+            Arrays.fill(nbytes[2], (byte)0);
+            Arrays.fill(nbytes[3], (byte)0);
+            Arrays.fill(nbytes[4], (byte)0);
+            Arrays.fill(nbytes[5], (byte)0);
+            Arrays.fill(nbytes[6], (byte)0);
+            Arrays.fill(nbytes[7], (byte)0);
+            DerValue val = DerValue.wrap(DerValue.tag_Sequence, out);
             key = val.toByteArray();
+            val.clear();
```

Comment:
Update RSAPrivateKeyImpl.java also for consistency sake?

Reply:
Good idea.